### ISSUE TITLE
spack install throws up Error with no explanation, output of --debug included

### ISSUE THREAD

**Issue by s-sajid-ali** [spack/spack#7935]
### Expected Result

spack install works as directed or gives an appropriate explanation as to why it fails

### Actual Result

spack throws up Error with no explanation at all. 

For example : 
```
[sajid@xrmlite ncurses-6.0]$ spack install python@3.6.5 %intel@19.0.0.046
==> bzip2 is already installed in /home/sajid/packages/spack/opt/spack/linux-centos7-x86_64/intel-19.0.0.046/bzip2-1.0.6-g6epdp7y4ina5z52yf5xpz74sngpkzyg
==> pkgconf is already installed in /home/sajid/packages/spack/opt/spack/linux-centos7-x86_64/intel-19.0.0.046/pkgconf-1.4.0-jjsb5qolv2aoepxe2zicbj2qnlu2jez3
==> Error:
```

### --debug output 
```
[sajid@xrmlite ncurses-6.0]$ spack --debug install python@3.6.5 %intel@19.0.0.046                                                                                                           
==> Reading config file /home/sajid/packages/spack/etc/spack/defaults/packages.yaml                                                                                                         
==> Reading config file /home/sajid/.spack/packages.yaml                                                                                                                                    
==> READ LOCK: /home/sajid/.spack/cache/providers/.builtin-index.yaml.lock[0:0] [Acquiring]                                                                                                 
==> READ LOCK: /home/sajid/.spack/cache/providers/.builtin-index.yaml.lock[0:0] [Acquired]                                                                                                  
==> READ LOCK: /home/sajid/.spack/cache/providers/.builtin-index.yaml.lock[0:0] [Released]                                                                                                  
==> Reading config file /home/sajid/.spack/linux/compilers.yaml                                                                                                                             
==> WRITE LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[6417505941298829586:1] [Acquiring]                                                                               
==> WRITE LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[6417505941298829586:1] [Acquired]                                                                                
==> READ LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/lock[0:0] [Acquiring]                                                                                                         
==> READ LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/lock[0:0] [Acquired]                                                                                                          
==> READ LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/lock[0:0] [Released]                                                                                                          
==> WRITE LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[6417505941298829586:1] [Released]                                                                                
==> link /home/sajid/packages/spack/var/spack/stage/python-3.6.5-wipr6tpoe2zckamozzji3ucpchg6cyae -> /home/sajid/Documents/build_temp/spack-stage/spack-stage-wvH0H9                        
==> READ LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[6417505941298829586:1] [Acquiring]                                                                                
==> READ LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[6417505941298829586:1] [Acquired]                                                                                 
==> READ LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[6417505941298829586:1] [Released]                                                                                 
==> Installing python dependencies                                                                                                                                                          
==> WRITE LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[2000857037867650311:1] [Acquiring]                                                                               
==> WRITE LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[2000857037867650311:1] [Acquired]                                                                                
==> WRITE LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[2000857037867650311:1] [Released]                                                                                
==> link /home/sajid/packages/spack/var/spack/stage/bzip2-1.0.6-g6epdp7y4ina5z52yf5xpz74sngpkzyg -> /home/sajid/Documents/build_temp/spack-stage/spack-stage-bzPW7F                         
==> READ LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[2000857037867650311:1] [Acquiring]                                                                                
==> READ LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[2000857037867650311:1] [Acquired]                                                                                 
==> bzip2 is already installed in /home/sajid/packages/spack/opt/spack/linux-centos7-x86_64/intel-19.0.0.046/bzip2-1.0.6-g6epdp7y4ina5z52yf5xpz74sngpkzyg                                   
==> READ LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[2000857037867650311:1] [Released]                                                                                 
==> WRITE LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[2680221637086232689:1] [Acquiring]                                                                               
==> WRITE LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[2680221637086232689:1] [Acquired]                                                                                
==> WRITE LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[2680221637086232689:1] [Released]                                                                                
==> link /home/sajid/packages/spack/var/spack/stage/pkgconf-1.4.0-jjsb5qolv2aoepxe2zicbj2qnlu2jez3 -> /home/sajid/Documents/build_temp/spack-stage/spack-stage-5bduq9                       
==> READ LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[2680221637086232689:1] [Acquiring]                                                                                
==> READ LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[2680221637086232689:1] [Acquired]                                                                                 
==> pkgconf is already installed in /home/sajid/packages/spack/opt/spack/linux-centos7-x86_64/intel-19.0.0.046/pkgconf-1.4.0-jjsb5qolv2aoepxe2zicbj2qnlu2jez3                               
==> READ LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[2680221637086232689:1] [Released]                                                                                 
==> WRITE LOCK: /home/sajid/packages/spack/opt/spack/.spack-db/prefix_lock[372156640868993180:1] [Acquiring]                                                                                
Traceback (most recent call last):                                                                                                                                                          
  File "/home/sajid/packages/spack/bin/spack", line 85, in <module>                                                                                                                         
    sys.exit(spack.main.main())                                                                                                                                                             
  File "/home/sajid/packages/spack/lib/spack/spack/main.py", line 577, in main                                                                                                              
    return _main(command, parser, args, unknown)                                                                                                                                            
  File "/home/sajid/packages/spack/lib/spack/spack/main.py", line 469, in _main                                                                                                             
    return _invoke_spack_command(command, parser, args, unknown_args)                                                                                                                       
  File "/home/sajid/packages/spack/lib/spack/spack/main.py", line 381, in _invoke_spack_command                                                                                             
    return_val = command(parser, args)                                                                                                                                                      
  File "/home/sajid/packages/spack/lib/spack/spack/cmd/install.py", line 256, in install                                                                                                    
    install_spec(args, kwargs, spec)                                                                                                                                                        
  File "/home/sajid/packages/spack/lib/spack/spack/cmd/install.py", line 150, in install_spec                                                                                               
    spec.package.do_install(**kwargs)                                                                                                                                                       
  File "/home/sajid/packages/spack/lib/spack/spack/package.py", line 1448, in do_install                                                                                                    
    **kwargs)                                                                                                                                                                               
  File "/home/sajid/packages/spack/lib/spack/spack/package.py", line 1411, in do_install                                                                                                    
    partial = self.check_for_unfinished_installation(keep_prefix, restage)                                                                                                                  
  File "/home/sajid/packages/spack/lib/spack/spack/package.py", line 1600, in check_for_unfinished_installation                                                                             
    with spack.store.db.prefix_write_lock(self.spec):                                                                                                                                       
  File "/usr/lib64/python2.7/contextlib.py", line 17, in __enter__                                                                                                                          
    return self.gen.next()                                                                                                                                                                  
  File "/home/sajid/packages/spack/lib/spack/spack/database.py", line 258, in prefix_write_lock                                                                                             
    prefix_lock.release_write()                                                                                                                                                             
  File "/home/sajid/packages/spack/lib/spack/llnl/util/lock.py", line 265, in release_write                                                                                                 
    assert self._writes > 0                                                                                                                                                                 
AssertionError                                                                                                                                                                              
```

### Information
system : linnux-centos7-x86_64
spack version : develop (up to date)

**Comment by adamjstewart** [comment]
@tgamblin knows the most about our locking logic.

**Comment by citibeth** [comment]
I'm having this problem too.  Try:
```
$ cd <spack top-level>
$ rm `find . *lock`
```

**Comment by s-sajid-ali** [comment]
Related to the above, now I repeatedly get this error : 
```
==> Do you want to proceed? [y/N] y
==> Error: Timed out waiting for lock.
```

**Comment by darmac** [comment]
@alalazo 
Any update of this issue? I have the same issue too, but not find how to resolve it yet.

**Comment by alalazo** [comment]
@darmac Locking changed completely after #13100 so I'm going to close this issue. Can you open a new issue for the problem you are encountering?

**Comment by darmac** [comment]
OK, I used a old version.
If I encounter the similar one, I'll report it.
